src/QUICConnection.ts

```diff
@@ -515,505 +549,762 @@ class QUICConnection extends EventTarget {
       // Note that a `0` timeout is still a valid timeout
       const timeout = this.conn.timeout();
       // If this is `null`, then quiche is requesting the timer to be cleaned up
       if (timeout == null) {
-        logger.debug('CLEANING UP TIMER');
+        if (this.conn.isTimedOut()) {
+          this.dispatchEvent(
+            new events.EventQUICConnectionError({
+              detail: new errors.ErrorQUICConnectionIdleTimeout()
+            })
+          );
+          this.dispatchEvent(
+            new events.EventQUICConnectionClose({
+              detail: {
+                type: 'timeout'
+              }
+            })
+          );
+        }
         return;
       }
       // Allow an extra 1ms for the delay to fully complete, so we can avoid a repeated 0ms delay
-      logger.debug(`Recreating timer with ${timeout + 1} delay`);
-      this.connTimeOutTimer = new Timer({
+      this.connTimeoutTimer = new Timer({
         delay: timeout + 1,
-        handler: connTimeOutHandler,
+        handler: connTimeoutHandler,
+        lazy: true,
       });
     };
     // Note that a `0` timeout is still a valid timeout
     const timeout = this.conn.timeout();
     // If this is `null`, then quiche is requesting the timer to be cleaned up
     if (timeout == null) {
-      // Clean up timer if it is running
-      if (
-        this.connTimeOutTimer != null &&
-        this.connTimeOutTimer.status === null
-      ) {
-        logger.debug('CLEANING UP TIMER');
-        this.connTimeOutTimer.cancel(timerCleanupReasonSymbol);
+      // Here we cancellation only matters if the timer status is `null` or settling
+      // If it is `null`, then the timer handler doesn't run
+      // If it is `settled`, then cancelling is a noop
+      // If it is `settling`, then cancelling only prevents it at the beginning
+      // Afterwards if it continues, it will continue to execute
+      this.connTimeoutTimer?.cancel();
+      if (this.conn.isTimedOut()) {
+        this.dispatchEvent(
+          new events.EventQUICConnectionError({
+            detail: new errors.ErrorQUICConnectionIdleTimeout()
+          })
+        );
+        this.dispatchEvent(
+          new events.EventQUICConnectionClose({
+            detail: {
+              type: 'timeout'
+            }
+          })
+        );
+      }
+      if (this.conn.isClosed()) {
+        this.resolveClosedP();
       }
       return;
     }
-    // If there was an existing timer, we cancel it and set a new one
+    // If there's no timer, create it
+    // If the timer is settled, create it
+    // If the timer is settling, do nothing (it will recreate itself)
+    // If the timer is null, reset it
+    // Plus 1 to the `timeout` to compensate for clock de-sync between JS and Rust
     if (
-      this.connTimeOutTimer != null &&
-      this.connTimeOutTimer.status === null
+      this.connTimeoutTimer == null ||
+      this.connTimeoutTimer.status === 'settled'
     ) {
```

Comment:
This condition doesn't re-create the timer after it has settled but not re-created itself. 

In that case the `connTimeoutTimer` exists but `connTimeoutTimer.status` is `null`. I'd expect it to be `'settled'` but I'm not sure why it's null. A bug in the timer implementation?

In any case, A quick fix solution is to delete the `this.connTimeoutTimer` when the timer has ended for whatever reason.

Reply:
This was done in 4e354752ae6281861c20ded1dc7334b0b0e276d4, we only needed to delete the timer when the timer is cancelled.

I think in the long run, the timer status should probably have: `settling | settled | cancelled | null`.

The `null` currently means both the timer hasn't executed yet and is still null when the timer is cancelled.

I might make the update first then apply it here, since in the long run, having a separate status for cancelled timer is a good thing.

Reply:
Actually I just ran tests in `js-timer`, and I have found that after cancellation, the timer is rejected, and after rejection, the timer is `settled`.

So `settled` can mean resolution or rejection, and cancellation and abortion both mean eventually rejection.

Thus it is not correct to assume that after cancellation the status will be `null`, it depends on how much time passes. Immediately afterwards, it's set to `null`, but eventually it will reject and become `settled`.

Reply:
Actually I just tested again:

```
      const t1 = new Timer(undefined, 100);
      console.log(t1.status);
      t1.cancel();
      console.log(t1.status);
```

And we can see that after `t1.cancel()` it is immediately `settled`.

Reply:
So that fix if `delete this.connTimeoutTimer` is not necessary.

Because if it were in fact cancelled, then `this.connTimeoutTimer.status` would be `settled`.

Reply:
I reverted that change in 4d3c9e2280e8e2cf8f2f2f48dacb039b874c2d35

Reply:
Can you re-check the condition you think is happneing here?

Reply:
So the fix is doing something... without it, indeed the QUICConnection is not stopping. However it should not be necessary. I still get errors somewhere, so I think something else is at fault here.

Reply:
I've actually re-enabled this change. However I don't think this is accurate. It need sto be checked again as why this is happening.

In general, it appears the code is not robust against immediate closures/destructions. It gets into a bad state, but not sure what that is yet.

The code does work fine as long as things are timed well.

Reply:
There are some concurrent close operations that don't seem to work well when things happen too quickly. Plus this comment still needs investigation since this shouldn't be a problem.

Reply:
This still shouldn't be necessary. Should recheck.

Reply:
Going to do this below.